### ISSUE TITLE
Discriminator array in subdoc

### ISSUE THREAD

**Issue by Tobjoern** [typegoose/typegoose#205]
## Versions

- typegoose: ^6.4.0
- mongoose: ^5.9.5

Hey!
I have a base class Event, and use discriminators to make two new classes ClickEvent, and VisitEvent, with an enum as a discriminant key.
I am able to save ClickEvent and VisitEvent successfully, all works great!
I now want to create a class EventsHolder, which holds an array of events (meaning VisitEvents or ClickEvents). However I am struggeling with this. When I save EventsHolder, with events attached, the db merely shows the properties originally defined in the Events class, not taking into account the ones defined via disciminators, which kind of makes sense.
I'm wondering if it is even possible to use discriminators in an array in a subdoc, or if I have to use references instead.

```ts
enum EventType {
    CLICK = 'VISIT',
    VISIT = 'VISIT'
}

@modelOptions({
    schemaOptions: {
        discriminatorKey: 'type',
    }
})
class Event {

    @prop({ required: true, enum: EventType })
    type!: EventType;

}

class ClickEvent extends Event {
    @prop({ required: true, default: 0 })
    public timesClicked!: number;
}

class VisitEvent extends Event {
    @prop({ required: true })
    public visitLength!: number;
}

const EventModel = getModelForClass(Event);
const ClickEventModel = getDiscriminatorModelForClass(EventModel, ClickEvent, EventType.CLICK);
const VisitEventModel = getDiscriminatorModelForClass(EventModel, VisitEvent, EventType.VISIT);

class EventsHolder {

    @arrayProp({ items: Event })
    events!: (VisitEvent | ClickEvent)[];

}

const EventsHolderModel = getModelForClass(EventsHolder)

    const evHolder = new EventsHolderModel()
    evHolder.events = [
        {
            type: EventType.CLICK,
            timesClicked: 4,
        },
        {
            type: EventType.VISIT,
            visitLength: 5
        },
    ]

    evHolder.save().catch(e => {
        console.log(e)
    }).then(() => {
        console.log('SAVED!')
    })
```
![image](https://user-images.githubusercontent.com/31477665/77157951-aabd4b80-6aa2-11ea-8b40-4977d7befd5f.png)

Thanks in advance!

**Comment by hasezoey** [comment]
i think this is a typo, but just to confirm: is it right that you use the same string for both values in the enum?

**Comment by hasezoey** [comment]
the thing with discriminator arrays is that typegoose currently cant support it, as you can see in the Pull Request #98 and #99, i cant continue on these / merge them, because i dont know what to actually archive there, and @B-Stefan is inactive (in this repo) since then

TL;DR: i dont fully understand discriminators and typegoose has problems supporting array / nested discriminators

---

btw, what do you want to archive here in the array? do you want them to be sub-documents or actual documents and the array just keeps ids, or find them [dynamically](https://typegoose.github.io/typegoose/docs/virtuals/#virtual-populate)?

**Comment by Tobjoern** [comment]
> the thing with discriminator arrays is that typegoose currently cant support it, as you can see in the Pull Request #98 and #99, i cant continue on these / merge them, because i dont know what to actually archive there, and @B-Stefan is inactive (in this repo) since then
> 
> TL;DR: i dont fully understand discriminators and typegoose has problems supporting array / nested discriminators
> 
> btw, what do you want to archive here in the array? do you want them to be sub-documents or actual documents and the array just keeps ids, or find them [dynamically](https://typegoose.github.io/typegoose/docs/virtuals/#virtual-populate)?

Yes, it is a typo, thanks for pointing that out!
I actually intended them to be subdocs, as this would be the most comvenient way, but it seems like this is not possible. I guess I will go with a reference to "Event". This should work, right?

**Comment by hasezoey** [comment]
i dont know much about discriminators so i guess they dont work as subdocs (at least not so easily)

> I guess I will go with a reference to "Event". This should work, right?

this should work, but i dont know the details

**Comment by hasezoey** [comment]
i will close this, because nested discriminators (and with that array-discriminators) got implemented (since about ~7.3)
